ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -776,6 +776,14 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = do
 
         return $ castPoint $ AF.headPoint newChain
       where
+        switchFollowerToFork curChain newChain ipoint =
+          let insertPoint h points = Set.insert (headerPoint h) points
+              oldPoints = AS.foldr insertPoint Set.empty
```

Comment:
Nice! So it seems that because we have a definition of the form:

```haskell
f x = let e = g x in \handle -> h handle
```

when we have an expression like `fmap (f y) handles`  `f y` is fully applied (?) causing `e` to be evaluated only once. If on the contrary we had written `f x handle = ...` `f y` won't be fully applied in the aforementioned `fmap` expression. I confirmed this by writing a toy program and profiling it, but I wonder if there is some documentation about this rule.

Reply:
I think it's just "standard" memoization of let-bindings?

Reply:
Thanks for profiling, @dnadales!